[Hardware][NV] Add support for ModelOpt static scaling checkpoints.

pavanimajety opened the pull request:
This change adds support for a new quantization class that has the
ability to load ModelOpt checkpoints through HF. It can be invoked via -

`llm = LLM(model=model_path, quantization="modelopt")`

The checkpoint needs a hf_quant_config.json for loading the right
quantiztion format.

<details>
<!-- inside this <details> section, markdown rendering does not work, so we use raw html here. -->
<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>


</details>

robertgshaw2-neuralmagic commented:
With the utilities that I made for fp8, all of the core functionality of fp8 is now refactored out of `fp8.py`.

So, this PR should not need to touch `fp8.py`. Instead, you can use the factored out utils with`ModelOptConfig` and `ModelOptFp8LinearMethod` that mirror the logic we used for AutoFP8

robertgshaw2-neuralmagic commented:
Utilities are in and this PR should now be unblocked

pavanimajety commented:
@robertgshaw2-neuralmagic @simon-mo @youkaichao Please review 🙏

robertgshaw2-neuralmagic commented:
@dsikka going to review

mgoin commented on vllm/model_executor/layers/quantization/modelopt.py:
```diff
@@ -0,0 +1,188 @@
+
+    @classmethod
+    def get_min_capability(cls) -> int:
+        return 80
```
Since this is only FP8 for the moment and FP8 Marlin isn't being used, we should restrict to GPUs with hardware support
```suggestion
        return 89
```

mgoin commented on vllm/model_executor/layers/quantization/modelopt.py:
```diff
@@ -0,0 +1,188 @@
+        quant_config = cls.get_from_keys(config, ["quantization"])
+        quant_method = quant_config["quant_algo"]
+        is_checkpoint_fp8_serialized = ("FP8" in quant_method)
+        activation_scheme = "static"
```
I believe you should check the config matches here

pavanimajety replied:
Config doesn't record "static" or "dynamic" with Model Optimizer config at the moment since it is only relevant for static quantization:
```
{
    "producer": {
        "name": "modelopt",
        "version": "0.17.0"
    },
    "quantization": {
        "quant_algo": "FP8",
        "kv_cache_quant_algo": null
    }
}

```

Removing references to "static" since it doesn't look relevant for this quantization class.

dsikka commented on vllm/model_executor/layers/quantization/modelopt.py:
```diff
@@ -0,0 +1,188 @@
+                "input_dim": 1,
+                "output_dim": 0,
+            },
+        )
```
Instead of using a raw torch.nn.Parameter, `ModelWeightParameter` should be used: https://github.com/vllm-project/vllm/blob/2188a60c7e0e5a414a87a4f0fd798333b2e0f625/vllm/model_executor/parameter.py#L166
```
weight = ModelWeightParameter(
data=torch.empty(
                output_size_per_partition,
                input_size_per_partition,
                dtype=weight_dtype,
            ),
input_dim=1,
output_dim=0,
weight_loader=weight_loader
)
```

This will eliminate the need for `set_weight_attrs`

See fp8 for an example: https://github.com/vllm-project/vllm/blob/2188a60c7e0e5a414a87a4f0fd798333b2e0f625/vllm/model_executor/layers/quantization/fp8.py#L155

dsikka commented on vllm/model_executor/layers/quantization/modelopt.py:
```diff
@@ -0,0 +1,188 @@
+        # Otherwise, wait until process_weights_after_loading.
+        if self.quant_config.is_checkpoint_fp8_serialized:
+            # WEIGHT SCALE
+            weight_scale = create_per_tensor_scale_param(
```
Similarly, these functions will be removed soon. `PerTensorScaleParameter` should be used instead.

Example: 

https://github.com/vllm-project/vllm/blob/2188a60c7e0e5a414a87a4f0fd798333b2e0f625/vllm/model_executor/layers/quantization/fp8.py#L177

pavanimajety committed:
[Hardware][NV] Add support for modelopt quantized weights as HF Ckpts

This change adds support for a new quantization class that has the
ability to load ModelOpt checkpoints through HF. It can be invoked via -

llm = LLM(model=model_path, quantization="modelopt")

The checkpoint needs a hf_quant_config.json for loading the right
quantiztion format. This config can be used to load the
nvidia/llama-3.1-*-FP8 models that are quantized using NVIDIA Model
Optimzier, and available on HF.

pavanimajety committed:
Fix formatting

pavanimajety commented:
/ready

mgoin commented on tests/models/test_modelopt.py:
```diff
@@ -0,0 +1,79 @@
+
+MAX_MODEL_LEN = 1024
+
+MODELS = ["nvidia/Llama-3.1-8B-Instruct-FP8"]
```
Just leaving a note that this model doesn't exist yet

mgoin reviewed (approved):
This looks good enough for now to land! Looking forward to a model checkpoint to test. You may want to merge with main to get the CI to pass

dsikka reviewed (approved):
LGTM
